### ISSUE TITLE
[Feature] allow complex key

### ISSUE THREAD

**Issue by avrahamcool** [tjmoses/crud-compare#5]
https://github.com/tjmoses/crud-compare/blob/af7a0e91af6983df0731603847c66811f4be9c3c/src/index.ts#L34

currently - the matching of object is based on a "key" property.
this has a limitation for object with complex keys (multiple properties) - or where any other logic should be applied to find the match.

I suggest to extend the "key" parameter, so it allow passing a Matcher function. (but also allow passing a simple string - so you don't create breaking changes)
`key: string | (left: any, right: any)=>bool`

the library will check if the given key is a string or a function.
if its a string - the library will continue just like always - look for a property with the name of the key.
but if a function is given - the library will apply the function on 2 items to conclude if they "match".

lets say we have a Person object, and the key is based on firstName + lastName - we could write something like:
```typescript
interface Person
{
   firstName: string;
   lastName: firstName;
   birthDate: Date;
   ...
}
compareObjectVals([arrBefore, arrAfter], (a, b) => a.firstName === b.firstName && a.lastName === b.lastName);
```

if it's fine by you - I can work on that a create a pull request.

**Comment by tjmoses** [comment]
Since this library is dealing with State, only comparison of keys within the object should ever be considered whether that's with natural, surrogate, or composite keys. I implemented basic keys within strings (natural and surrogate), but do think composite keys would be nice. Additional checks would be introducing side-effects, and limiting those is best.

I don't think a matcher function is beneficial to everyone in the sense that most would not provide this and if they are writing a basic matcher function, they probably already know what they want to compare and won't use this library. Although, you bring up a valid point in that composite keys should be implemented. 

I would be open to adding that in with an additional array argument of keys vs the single that exists. Also, from your PR using indexOf is faster than several other methods and why I used it. I mean I can make it extremely short with 100% ES6+ code, but I was concerned with performance and speed. Anyways, if you want to implement something that can take a given list of composite keys to check I'm game. Thanks for your input!
